gitignore globs are read from enclosing directories, so tree is not copied

I'm trying to run `cargo-mutants` for my ubi crate (https://github.com/houseabsolute/ubi/) like this:

```
cargo mutants -p ubi -L trace
```

This gives me the following output:

```
DEBUG output: {"root":"/home/autarch/projects/ubi/Cargo.toml"}
DEBUG Found workspace root manifest cargo_toml_path="/home/autarch/projects/ubi/Cargo.toml"
DEBUG Find root files manifest_path="/home/autarch/projects/ubi/Cargo.toml"
DEBUG Found workspace root workspace_root="/home/autarch/projects/ubi"
DEBUG package{name=ubi}: walk package manifest_path=/home/autarch/projects/ubi/ubi/Cargo.toml
DEBUG package{name=ubi}: found mutation target ubi/src/lib.rs of kind ["lib"]
DEBUG package{name=ubi}: skipping target "build-script-build" of kinds ["custom-build"]
DEBUG package{name=ubi-cli}: walk package manifest_path=/home/autarch/projects/ubi/ubi-cli/Cargo.toml
DEBUG package{name=ubi-cli}: found mutation target ubi-cli/src/main.rs of kind ["bin"]
DEBUG package{name=ubi-cli}: skipping target "ubi" of kinds ["test"]
DEBUG config=Config { cap_lints: false, error_values: [], examine_globs: [], exclude_globs: [], exclude_re: [], examine_re: [], additional_cargo_args: [], ad}
DEBUG args.features=Features { features: [], no_default_features: false, all_features: false }
DEBUG options=Options { baseline: Run, cap_lints: false, check_only: false, gitignore: true, in_place: false, jobserver: true, jobserver_tasks: None, leak_di}
DEBUG source_file{path="ubi/src/lib.rs"}: visit source file
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=93 name="project"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=101 name="tag"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path { }
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=113 name="url"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path { }
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=120 name="install_dir"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path:}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=130 name="matching"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pa}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=139 name="exe"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path { }
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=147 name="github_token"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=155 name="platform"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pa}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=164 name="is_musl"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pat}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=172 name="github_api_url_base"}: Return type is not recognized, trying Default type_=Type::Path { qself: Non}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=186 name="build"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path }
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=186 name="build"}:binary{line=187}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=186 name="build"}:binary{line=190}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=186 name="build"}:binary{line=190}:binary{line=190}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=186 name="build"}:binary{line=207}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=186 name="build"}:binary{line=207}:binary{line=207}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=232 name="platform_is_musl"}: visit fn
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=232 name="platform_is_musl"}:binary{line=233}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=232 name="platform_is_musl"}:binary{line=247}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=296 name="parse_project_name"}:binary{line=313}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=296 name="parse_project_name"}:binary{line=313}:binary{line=313}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=296 name="parse_project_name"}:binary{line=313}:binary{line=313}:binary{line=313}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=324 name="exe_name"}:binary{line=332}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=343 name="install_path"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=357 name="reqwest_client"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, pa}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=406 name="asset"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path }
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=416 name="download_asset"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, pa}
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=416 name="download_asset"}:binary{line=424}: visit binary operator
TRACE source_file{path="ubi/src/lib.rs"}:fn{line=465 name="init_logger"}: visit fn
DEBUG source_file{path="ubi/src/lib.rs"}:fn{line=465 name="init_logger"}: Attribute is not in conventional form; skipped err=Error("expected `,`") attr=Attri}
TRACE source_file{path="ubi/src/lib.rs"}:mod{line=495 mod_name="test"}: mod excluded by attrs
TRACE found submodule in /home/autarch/projects/ubi/ubi/src/arch.rs
TRACE found submodule in /home/autarch/projects/ubi/ubi/src/extension.rs
TRACE found submodule in /home/autarch/projects/ubi/ubi/src/fetcher.rs
TRACE found submodule in /home/autarch/projects/ubi/ubi/src/installer.rs
TRACE found submodule in /home/autarch/projects/ubi/ubi/src/os.rs
TRACE found submodule in /home/autarch/projects/ubi/ubi/src/picker.rs
TRACE found submodule in /home/autarch/projects/ubi/ubi/src/release.rs
DEBUG source_file{path="ubi/src/arch.rs"}: visit source file
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=7 name="macos_aarch64_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=7 name="macos_aarch64_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, p}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=35 name="aarch64_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=35 name="aarch64_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: }
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=57 name="arm_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=57 name="arm_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=75 name="mipsle_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=75 name="mipsle_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: P}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=93 name="mips_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=93 name="mips_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pat}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=111 name="mips64le_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=111 name="mips64le_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=129 name="mips64_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=129 name="mips64_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: }
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=147 name="ppc32_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=147 name="ppc32_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: P}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=169 name="ppc64_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=169 name="ppc64_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: P}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=202 name="ppc64le_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=202 name="ppc64le_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path:}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=235 name="riscv64_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=235 name="riscv64_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path:}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=253 name="s390x_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=253 name="s390x_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: P}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=271 name="sparc64_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=271 name="sparc64_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path:}
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=289 name="x86_32_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=289 name="x86_32_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: }
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=315 name="x86_64_re"}: visit fn
TRACE source_file{path="ubi/src/arch.rs"}:fn{line=315 name="x86_64_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: }
DEBUG source_file{path="ubi/src/extension.rs"}: visit source file
TRACE source_file{path="ubi/src/extension.rs"}:fn{line=56 name="from_path"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, pa}
TRACE source_file{path="ubi/src/extension.rs"}:fn{line=90 name="extension_is_part_of_version"}: visit fn
TRACE source_file{path="ubi/src/extension.rs"}:fn{line=90 name="extension_is_part_of_version"}:binary{line=109}: visit binary operator
TRACE source_file{path="ubi/src/extension.rs"}:fn{line=112 name="extension_is_platform"}: visit fn
TRACE source_file{path="ubi/src/extension.rs"}:mod{line=127 mod_name="test"}: mod excluded by attrs
DEBUG source_file{path="ubi/src/fetcher.rs"}: visit source file
TRACE source_file{path="ubi/src/fetcher.rs"}:fn{line=31 name="fetch_assets"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, p}
TRACE source_file{path="ubi/src/fetcher.rs"}:fn{line=42 name="release_info"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, p}
DEBUG source_file{path="ubi/src/installer.rs"}: visit source file
TRACE source_file{path="ubi/src/installer.rs"}:fn{line=90 name="extract_tarball"}:unary{line=100}: visit unary operator
TRACE source_file{path="ubi/src/installer.rs"}:fn{line=90 name="extract_tarball"}:binary{line=106}: visit binary operator
TRACE source_file{path="ubi/src/installer.rs"}:fn{line=171 name="tar_reader_for"}: visit fn
TRACE source_file{path="ubi/src/installer.rs"}:fn{line=171 name="tar_reader_for"}: Return type is not recognized, trying Default type_=Type::TraitObject { dy}
TRACE source_file{path="ubi/src/installer.rs"}:fn{line=194 name="open_file"}: visit fn
TRACE source_file{path="ubi/src/installer.rs"}:fn{line=194 name="open_file"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, p}
TRACE source_file{path="ubi/src/installer.rs"}:mod{line=199 mod_name="tests"}: mod excluded by attrs
DEBUG source_file{path="ubi/src/os.rs"}: visit source file
TRACE source_file{path="ubi/src/os.rs"}:fn{line=5 name="freebsd_re"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=5 name="freebsd_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pat}
TRACE source_file{path="ubi/src/os.rs"}:fn{line=9 name="fuchsia"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=9 name="fuchsia"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path {}
TRACE source_file{path="ubi/src/os.rs"}:fn{line=13 name="illumos_re"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=13 name="illumos_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pa}
TRACE source_file{path="ubi/src/os.rs"}:fn{line=17 name="linux_re"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=17 name="linux_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path}
TRACE source_file{path="ubi/src/os.rs"}:fn{line=21 name="macos_re"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=21 name="macos_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Path}
TRACE source_file{path="ubi/src/os.rs"}:fn{line=25 name="netbsd_re"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=25 name="netbsd_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pat}
TRACE source_file{path="ubi/src/os.rs"}:fn{line=29 name="solaris_re"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=29 name="solaris_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pa}
TRACE source_file{path="ubi/src/os.rs"}:fn{line=33 name="windows_re"}: visit fn
TRACE source_file{path="ubi/src/os.rs"}:fn{line=33 name="windows_re"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path: Pa}
DEBUG source_file{path="ubi/src/picker.rs"}: visit source file
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=34 name="pick_asset"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=34 name="pick_asset"}:binary{line=47}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=86 name="os_matches"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, path}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=108 name="arch_matches"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, p}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=108 name="arch_matches"}:binary{line=116}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=158 name="libc_matches"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, p}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=158 name="libc_matches"}:unary{line=159}: visit unary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=158 name="libc_matches"}:binary{line=168}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=183 name="libc_name"}:binary{line=186}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=193 name="pick_asset_from_matches"}: Return type is not recognized, trying Default type_=Type::Path { qse}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=193 name="pick_asset_from_matches"}:binary{line=194}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=193 name="pick_asset_from_matches"}:binary{line=206}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=230 name="maybe_filter_for_64_bit_arch"}: Return type is not recognized, trying Default type_=Type::Path }
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=230 name="maybe_filter_for_64_bit_arch"}:unary{line=231}: visit unary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=230 name="maybe_filter_for_64_bit_arch"}:unary{line=250}: visit unary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=266 name="maybe_filter_for_matching_string"}: Return type is not recognized, trying Default type_=Type::P}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=266 name="maybe_filter_for_matching_string"}: Return type is not recognized, trying Default type_=Type::P}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=290 name="maybe_pick_asset_for_macos_arm"}: Return type is not recognized, trying Default type_=Type::Pat}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=290 name="maybe_pick_asset_for_macos_arm"}: Return type is not recognized, trying Default type_=Type::Pat}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=290 name="maybe_pick_asset_for_macos_arm"}:unary{line=294}: visit unary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=315 name="maybe_filter_for_musl_platform"}: Return type is not recognized, trying Default type_=Type::Pat}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=315 name="maybe_filter_for_musl_platform"}:unary{line=316}: visit unary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=315 name="maybe_filter_for_musl_platform"}:unary{line=326}: visit unary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=342 name="os_matcher"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, pat}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=372 name="arch_matcher"}: Return type is not recognized, trying Default type_=Type::Path { qself: None, p}
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=404 name="running_on_macos_arm"}:binary{line=405}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=404 name="running_on_macos_arm"}:binary{line=405}:binary{line=405}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:fn{line=404 name="running_on_macos_arm"}:binary{line=405}:binary{line=405}: visit binary operator
TRACE source_file{path="ubi/src/picker.rs"}:mod{line=410 mod_name="test"}: mod excluded by attrs
DEBUG source_file{path="ubi/src/release.rs"}: visit source file
Found 205 mutants to test
DEBUG glob converted to regex: Glob { glob: "**/*", re: "(?-u)^(?:/?|.*/)[^/]*$", opts: GlobOptions { case_insensitive: false, literal_separator: true, backs 
DEBUG built glob set; 0 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 1 regexes    
DEBUG built glob set; 4 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 0 regexes    
DEBUG ignoring /home/autarch/projects/ubi/package-lock.json: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/projects/ubi/.gitignore"), origina 
DEBUG ignoring /home/autarch/projects/ubi/.github: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/*", 
DEBUG ignoring /home/autarch/projects/ubi/ubi-cli: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/*", 
DEBUG ignoring /home/autarch/projects/ubi/LICENSE-MIT: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "** 
DEBUG ignoring /home/autarch/projects/ubi/Cargo.lock: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/ 
DEBUG ignoring /home/autarch/projects/ubi/mutants.out: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "** 
DEBUG ignoring /home/autarch/projects/ubi/do-release.sh: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: " 
DEBUG ignoring /home/autarch/projects/ubi/Cross.toml: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/ 
DEBUG ignoring /home/autarch/projects/ubi/.git: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/*", is 
DEBUG ignoring /home/autarch/projects/ubi/LICENSE-APACHE: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual:  
DEBUG ignoring /home/autarch/projects/ubi/release.toml: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "* 
DEBUG ignoring /home/autarch/projects/ubi/ubi: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/*", is_ 
DEBUG ignoring /home/autarch/projects/ubi/.gitignore: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/ 
DEBUG ignoring /home/autarch/projects/ubi/.typos.toml: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "** 
DEBUG ignoring /home/autarch/projects/ubi/node_modules: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/projects/ubi/.gitignore"), original: "/ 
DEBUG ignoring /home/autarch/projects/ubi/dev: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/*", is_ 
DEBUG ignoring /home/autarch/projects/ubi/Cargo.toml: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/ 
DEBUG ignoring /home/autarch/projects/ubi/bootstrap: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/* 
DEBUG ignoring /home/autarch/projects/ubi/precious.toml: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: " 
DEBUG ignoring /home/autarch/projects/ubi/Changes.md: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/ 
DEBUG ignoring /home/autarch/projects/ubi/rustfmt.toml: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "* 
DEBUG ignoring /home/autarch/projects/ubi/target: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/projects/ubi/.gitignore"), original: "/target 
DEBUG ignoring /home/autarch/projects/ubi/README.md: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/* 
DEBUG ignoring /home/autarch/projects/ubi/package.json: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/projects/ubi/.gitignore"), original: "/ 
DEBUG ignoring /home/autarch/projects/ubi/git: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("/home/autarch/.gitignore"), original: "*", actual: "**/*", is_ 
DEBUG Copied source tree total_bytes=0 total_files=0 temp_dir="/tmp/cargo-mutants-ubi-WI5h83.tmp"
Error: read manifest

Caused by:
    No such file or directory (os error 2)
```

If I pass `--leak-dirs`, I note that the directory in `/tmp` doesn't seem to exist. I tried setting `TMPDIR=$HOME/tmp` and that led to the same error, which no temp directory being created.

Hi, thanks for the report. 

I'll try to reproduce this.

On a first attempt, I see the ubi tests fail, perhaps because the integration tests need a GitHub API token?

```
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running `/tmp/cargo-mutants-ubi-pXw0V7.tmp/target/debug/deps/ubi-2346e85f4aea3af0`

running 1 test
test integration_tests ... FAILED

failures:

---- integration_tests stdout ----
running cargo build
running /tmp/cargo-mutants-ubi-pXw0V7.tmp/ubi-cli/../target/debug/ubi --project houseabsolute/precious
Error: ran /tmp/cargo-mutants-ubi-pXw0V7.tmp/ubi-cli/../target/debug/ubi --project houseabsolute/precious and got non-zero exit code: 1
Stderr:
[ubi][ERROR] HTTP status client error (401 Unauthorized) for url (https://api.github.com/repos/houseabsolute/precious/releases/latest)



failures:
    integration_tests

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 6.50s

```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "cargo-mutants"
-version = "24.11.1"
+version = "24.11.2"
 edition = "2021"
 authors = ["Martin Pool"]
 license = "MIT"

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,5 +1,11 @@
 # cargo-mutants changelog
 
+## Unreleased
+
+- Changed: `.gitignore` (and other git ignore files) are only consulted when copying the tree if it is contained within a directory with a `.git` directory.
+
+- Fixed: `.gitignore` files above the git root directory are no longer read. In particular this fixes the problem where `.gitignore *` in the home directory would prevent copying any source trees.
+
 ## 24.11.1
 
 - Changed: The arguments of calls to functions or methods named `with_capacity` are not mutated by default. This can be turned off with `--skip-calls-defaults=false` on the command line or `skip_calls_defaults = false` in `.cargo/mutants.toml`.

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,6 +1,6 @@
 # cargo-mutants changelog
 
-## Unreleased
+## 24.11.2
 
 - Changed: `.gitignore` (and other git ignore files) are only consulted when copying the tree if it is contained within a directory with a `.git` directory.
 

--- a/book/src/build-dirs.md
+++ b/book/src/build-dirs.md
@@ -17,4 +17,8 @@ Files or directories matching these patterns are not copied:
 
 From 23.11.2, by default, cargo-mutants will not copy files that are excluded by gitignore patterns, to make copying faster in large trees.
 
-This behavior can be turned off with `--gitignore=false`.
+gitignore filtering is only used within trees containing a `.git` directory.
+
+The filter, based on the [`ignore` crate](https://docs.rs/ignore/), also respects global git ignore configuration in the home directory, as well as `.gitignore` files within the tree.
+
+This behavior can be turned off with `--gitignore=false`, causing ignored files to be copied.

--- a/src/copy_tree.rs
+++ b/src/copy_tree.rs
@@ -57,16 +57,17 @@ pub fn copy_tree(
         .try_into()
         .context("Convert path to UTF-8")?;
     console.start_copy(dest);
-    for entry in WalkBuilder::new(from_path)
+    let mut walk_builder = WalkBuilder::new(from_path);
+    walk_builder
         .standard_filters(gitignore)
         .hidden(false) // copy hidden files
         .ignore(false) // don't use .ignore
         .require_git(true) // stop at git root; only read gitignore files inside git trees
         .filter_entry(|entry| {
             !SOURCE_EXCLUDE.contains(&entry.file_name().to_string_lossy().as_ref())
-        })
-        .build()
-    {
+        });
+    debug!(?walk_builder);
+    for entry in walk_builder.build() {
         check_interrupted()?;
         let entry = entry?;
         let relative_path = entry

--- a/src/copy_tree.rs
+++ b/src/copy_tree.rs
@@ -2,8 +2,6 @@
 
 //! Copy a source tree, with some exclusions, to a new temporary directory.
 
-use std::fs::FileType;
-
 use anyhow::Context;
 use camino::{Utf8Path, Utf8PathBuf};
 use ignore::WalkBuilder;
@@ -11,9 +9,17 @@ use path_slash::PathExt;
 use tempfile::TempDir;
 use tracing::{debug, warn};
 
-use crate::check_interrupted;
-use crate::Console;
-use crate::Result;
+use crate::{check_interrupted, Console, Result};
+
+#[cfg(unix)]
+mod unix;
+#[cfg(unix)]
+use unix::copy_symlink;
+
+#[cfg(windows)]
+mod windows;
+#[cfg(windows)]
+use windows::copy_symlink;
 
 /// Filenames excluded from being copied with the source.
 static SOURCE_EXCLUDE: &[&str] = &[
@@ -104,31 +110,4 @@ pub fn copy_tree(
     console.finish_copy(dest);
     debug!(?total_bytes, ?total_files, temp_dir = ?temp_dir.path(), "Copied source tree");
     Ok(temp_dir)
-}
-
-#[cfg(unix)]
-fn copy_symlink(_ft: FileType, src_path: &Utf8Path, dest_path: &Utf8Path) -> Result<()> {
-    let link_target = std::fs::read_link(src_path)
-        .with_context(|| format!("Failed to read link {src_path:?}"))?;
-    std::os::unix::fs::symlink(link_target, dest_path)
-        .with_context(|| format!("Failed to create symlink {dest_path:?}",))?;
-    Ok(())
-}
-
-#[cfg(windows)]
-#[mutants::skip] // Mutant tests run on Linux
-fn copy_symlink(ft: FileType, src_path: &Utf8Path, dest_path: &Utf8Path) -> Result<()> {
-    use std::os::windows::fs::FileTypeExt;
-    let link_target =
-        std::fs::read_link(src_path).with_context(|| format!("read link {src_path:?}"))?;
-    if ft.is_symlink_dir() {
-        std::os::windows::fs::symlink_dir(link_target, dest_path)
-            .with_context(|| format!("create symlink {dest_path:?}"))?;
-    } else if ft.is_symlink_file() {
-        std::os::windows::fs::symlink_file(link_target, dest_path)
-            .with_context(|| format!("create symlink {dest_path:?}"))?;
-    } else {
-        anyhow::bail!("Unknown symlink type: {:?}", ft);
-    }
-    Ok(())
 }

--- a/src/copy_tree.rs
+++ b/src/copy_tree.rs
@@ -59,9 +59,9 @@ pub fn copy_tree(
     console.start_copy(dest);
     for entry in WalkBuilder::new(from_path)
         .standard_filters(gitignore)
-        .hidden(false)
-        .ignore(false)
-        .require_git(false)
+        .hidden(false) // copy hidden files
+        .ignore(false) // don't use .ignore
+        .require_git(true) // stop at git root; only read gitignore files inside git trees
         .filter_entry(|entry| {
             !SOURCE_EXCLUDE.contains(&entry.file_name().to_string_lossy().as_ref())
         })
@@ -110,4 +110,40 @@ pub fn copy_tree(
     console.finish_copy(dest);
     debug!(?total_bytes, ?total_files, temp_dir = ?temp_dir.path(), "Copied source tree");
     Ok(temp_dir)
+}
+
+#[cfg(test)]
+mod test {
+    use std::fs::{create_dir, write};
+
+    use camino::Utf8PathBuf;
+    use tempfile::TempDir;
+
+    use crate::console::Console;
+    use crate::Result;
+
+    use super::copy_tree;
+
+    /// Test for regression of <https://github.com/sourcefrog/cargo-mutants/issues/450>
+    #[test]
+    fn copy_tree_with_parent_ignoring_star() -> Result<()> {
+        let tmp_dir = TempDir::new().unwrap();
+        let tmp = tmp_dir.path();
+        write(tmp.join(".gitignore"), "*\n")?;
+
+        let a = Utf8PathBuf::try_from(tmp.join("a")).unwrap();
+        create_dir(&a)?;
+        write(a.join("Cargo.toml"), "[package]\nname = a")?;
+        let src = a.join("src");
+        create_dir(&src)?;
+        write(src.join("main.rs"), "fn main() {}")?;
+
+        let dest_tmpdir = copy_tree(&a, "a", true, &Console::new())?;
+        let dest = dest_tmpdir.path();
+        assert!(dest.join("Cargo.toml").is_file());
+        assert!(dest.join("src").is_dir());
+        assert!(dest.join("src/main.rs").is_file());
+
+        Ok(())
+    }
 }

--- a/src/copy_tree/unix.rs
+++ b/src/copy_tree/unix.rs
@@ -1,0 +1,14 @@
+use std::fs::FileType;
+
+use anyhow::Context;
+use camino::Utf8Path;
+
+use crate::Result;
+
+pub(super) fn copy_symlink(_ft: FileType, src_path: &Utf8Path, dest_path: &Utf8Path) -> Result<()> {
+    let link_target = std::fs::read_link(src_path)
+        .with_context(|| format!("Failed to read link {src_path:?}"))?;
+    std::os::unix::fs::symlink(link_target, dest_path)
+        .with_context(|| format!("Failed to create symlink {dest_path:?}",))?;
+    Ok(())
+}

--- a/src/copy_tree/windows.rs
+++ b/src/copy_tree/windows.rs
@@ -1,0 +1,22 @@
+use std::fs::FileType;
+use std::os::windows::fs::FileTypeExt;
+
+use anyhow::Context;
+use camino::Utf8Path;
+
+use crate::Result;
+#[mutants::skip] // Mutant tests run on Linux
+fn copy_symlink(ft: FileType, src_path: &Utf8Path, dest_path: &Utf8Path) -> Result<()> {
+    let link_target =
+        std::fs::read_link(src_path).with_context(|| format!("read link {src_path:?}"))?;
+    if ft.is_symlink_dir() {
+        std::os::windows::fs::symlink_dir(link_target, dest_path)
+            .with_context(|| format!("create symlink {dest_path:?}"))?;
+    } else if ft.is_symlink_file() {
+        std::os::windows::fs::symlink_file(link_target, dest_path)
+            .with_context(|| format!("create symlink {dest_path:?}"))?;
+    } else {
+        anyhow::bail!("Unknown symlink type: {:?}", ft);
+    }
+    Ok(())
+}

--- a/src/copy_tree/windows.rs
+++ b/src/copy_tree/windows.rs
@@ -6,7 +6,7 @@ use camino::Utf8Path;
 
 use crate::Result;
 #[mutants::skip] // Mutant tests run on Linux
-fn copy_symlink(ft: FileType, src_path: &Utf8Path, dest_path: &Utf8Path) -> Result<()> {
+pub(super) fn copy_symlink(ft: FileType, src_path: &Utf8Path, dest_path: &Utf8Path) -> Result<()> {
     let link_target =
         std::fs::read_link(src_path).with_context(|| format!("read link {src_path:?}"))?;
     if ft.is_symlink_dir() {

--- a/src/manifest.rs
+++ b/src/manifest.rs
@@ -1,11 +1,11 @@
-// Copyright 2022-2023 Martin Pool.
+// Copyright 2022-2024 Martin Pool.
 
 //! Manipulate Cargo manifest and config files.
 //!
 //! In particular, when the tree is copied we have to fix up relative paths, so
 //! that they still work from the new location of the scratch directory.
 
-use std::fs;
+use std::fs::{read_to_string, write};
 
 use anyhow::Context;
 use camino::Utf8Path;
@@ -19,11 +19,15 @@ use crate::Result;
 /// `manifest_source_dir` is the directory originally containing the manifest, from
 /// which the absolute paths are calculated.
 pub fn fix_manifest(manifest_scratch_path: &Utf8Path, source_dir: &Utf8Path) -> Result<()> {
-    let toml_str = fs::read_to_string(manifest_scratch_path).context("read manifest")?;
+    let toml_str = read_to_string(manifest_scratch_path).with_context(|| {
+        format!("failed to read manifest from build directory: {manifest_scratch_path}")
+    })?;
     if let Some(changed_toml) = fix_manifest_toml(&toml_str, source_dir)? {
         let toml_str =
             toml::to_string_pretty(&changed_toml).context("serialize changed manifest")?;
-        fs::write(manifest_scratch_path, toml_str.as_bytes()).context("write manifest")?;
+        write(manifest_scratch_path, toml_str.as_bytes()).with_context(|| {
+            format!("Failed to write fixed manifest to {manifest_scratch_path}")
+        })?;
     }
     Ok(())
 }
@@ -101,9 +105,9 @@ fn fix_dependency_table(dependencies: &mut toml::Value, manifest_source_dir: &Ut
 pub fn fix_cargo_config(build_path: &Utf8Path, source_path: &Utf8Path) -> Result<()> {
     let config_path = build_path.join(".cargo/config.toml");
     if config_path.exists() {
-        let toml_str = fs::read_to_string(&config_path).context("read .cargo/config.toml")?;
+        let toml_str = read_to_string(&config_path).context("read .cargo/config.toml")?;
         if let Some(changed_toml) = fix_cargo_config_toml(&toml_str, source_path)? {
-            fs::write(build_path.join(&config_path), changed_toml.as_bytes())
+            write(build_path.join(&config_path), changed_toml.as_bytes())
                 .context("write .cargo/config.toml")?;
         }
     }

--- a/tests/build_dir.rs
+++ b/tests/build_dir.rs
@@ -1,6 +1,6 @@
-// Copyright 2023 Martin Pool
+// Copyright 2023-2024 Martin Pool
 
-use std::fs::write;
+use std::fs::{create_dir, write};
 
 mod util;
 use util::{copy_of_testdata, run};
@@ -10,6 +10,9 @@ fn gitignore_respected_in_copy_by_default() {
     // Make a tree with a (dumb) gitignore that excludes the source file; when you copy it
     // to a build directory, the source file should not be there and so the check will fail.
     let tmp = copy_of_testdata("factorial");
+    // There must be something that looks like a `.git` dir, otherwise we don't read
+    // `.gitignore` files.
+    create_dir(tmp.path().join(".git")).unwrap();
     write(tmp.path().join(".gitignore"), b"src\n").unwrap();
     run()
         .args(["mutants", "--check", "-d"])

